pallets/providers/src/utils.rs

```diff
@@ -2267,4 +2384,186 @@ where
+            if !meter.can_consume(minimum_required_weight_processing_expired_items) {
+                return;
+            }
+
+            // Provider top up expirations expire based on the relay chain block number, not the local block number.
+            // TODO: use next starting block to clean up instead of current block number here
+            let relay_chain_block_number = RelayBlockGetter::<T>::current_block_number();
+
+            // Remove expired move bucket requests if any existed and process them.
+            let mut provider_top_up_expirations =
+                ProviderTopUpExpirations::<T>::take(&relay_chain_block_number);
+            meter.consume(minimum_required_weight_processing_expired_items);
+
+            // TODO: After benchmarking, we should check before this loop that there is enough remaining weight to
+            // TODO: process all the expired move bucket requests. If not, we should return early.
+            while let Some(provider_id) = provider_top_up_expirations.pop() {
+                Self::process_expired_provider_top_up_period(provider_id, meter);
+            }
+
+            // If there are remaining items which were not processed, put them back in storage
+            if !provider_top_up_expirations.is_empty() {
+                ProviderTopUpExpirations::<T>::insert(
+                    &relay_chain_block_number,
+                    provider_top_up_expirations,
+                );
+                meter.consume(db_weight.writes(1));
+            }
+        }
+
+        fn process_expired_provider_top_up_period(
+            provider_id: ProviderIdFor<T>,
+            meter: &mut WeightMeter,
+        ) {
+            let db_weight = T::DbWeight::get();
+            let potential_weight = db_weight.reads_writes(0, 2);
+
+            if !meter.can_consume(potential_weight) {
+                return;
+            }
+
+            // Clear awaiting top up storage
+            let maybe_awaiting_top_up = AwaitingTopUpFromProviders::<T>::take(&provider_id);
+
+            // Mark the provider as insolvent if it was awaiting a top up
+            // If the provider was not awaiting a top up, it means they already topped up either via an
+            // automatic top up or a manual top up.
+            let current_tick =
+                <T::PaymentStreams as shp_traits::PaymentStreamsInterface>::current_tick();
+            if maybe_awaiting_top_up.is_some() {
+                InsolventProviders::<T>::insert(provider_id, current_tick);
+
+                Self::deposit_event(Event::ProviderInsolvent { provider_id });
+
+                let account_id = if let Some(bsp) = BackupStorageProviders::<T>::get(&provider_id) {
+                    bsp.owner_account
+                } else if let Some(msp) = MainStorageProviders::<T>::get(&provider_id) {
+                    msp.owner_account
+                } else {
+                    log::warn!(
```

Comment:
Debatable whether we should keep this log in the runtime. But if we do, I think this should be an error, as it would be a serious state inconsistency.

Reply:
Let's make this an error log or just remove it.

Reply:
changed to error